### PR TITLE
Expand linear basis to support variable LJ sigma parameters

### PR DESCRIPTION
**Pull request by maxentile**
Continuation of https://github.com/proteneer/timemachine/pull/685  -- which only handled variable LJ **epsilon** parameters -- to handle also variable LJ **sigma** parameters.

Uses an approach (from eq. C.1. of Levi Naden's thesis) that:
* Increases the size of the basis expansion (from 1 to 20 numbers per ligand atom)
* Does not scale with the number of environment atom types
* Does not require switching [from arithmetic mixing rule](https://github.com/proteneer/timemachine/blob/9d1c9baa0db7c9313ed998d87e7b56fddfc3a417/timemachine/potentials/nonbonded.py#L31) to geometric mixing rule for sigma parameters

~TODO: update tests~

### CODE REVIEW

**Comment by maxentile** [comment]
For future reference, this optimization can also be re-discovered by applying the sympy functions `expand` and `factor` to the following function:

```python
from sympy import summation, symbols, IndexedBase, Idx

# variable atom sigma, epsilon
sig = symbols('sigma')
eps = symbols('epsilon')

# environment atoms i = 0 ... K
i = symbols('i', cls=Idx)
K = symbols('K', type=int)

eps_env = IndexedBase('epsilon^env')
sig_env = IndexedBase('sigma^env')

# distance between variable atom and each environment atom
r_env = IndexedBase('r^env')


def u(sig, eps):
    combined_sig = sig + sig_env[i]

    sig_r_12 = (combined_sig / r_env[i])**12
    sig_r_6 = (combined_sig / r_env[i])**6

    energy = eps * eps_env[i] * (sig_r_12 - sig_r_6)

    return summation(energy, (i, 0, K))
```

(more details in this notebook: https://gist.github.com/maxentile/c9cc8d8cff5d5758919c363bdb8744be )

I wasn't quickly able to reduce the number of basis functions from the current (12 + 1) + (6 + 1) = 20 with the help of sympy, but maybe further simplifications can be found in the future.

File: timemachine/potentials/nonbonded.py [hunk 1]
```diff
@@ -565,31 +648,30 @@ def lj_prefactors_on_traj(traj, boxes, sigmas, epsilons, ligand_indices, env_ind
...
     eps_env = epsilons[env_indices]
     sig_env = sigmas[env_indices]
 
     def f_snapshot(coords, box):
         x_ligand = coords[ligand_indices]
         x_env = coords[env_indices]
-        return lj_prefactors_on_snapshot(x_ligand, x_env, sig_ligand, sig_env, eps_env, box, cutoff)
+        return lj_prefactors_on_snapshot(x_ligand, x_env, sig_env, eps_env, box, cutoff)
 
     return vmap(f_snapshot)(traj, boxes)
```
Comment (jkausrelay) [inline timemachine/potentials/nonbonded.py:659]: @maxentile This is the part that can use a lot of memory. The intermediate computation of `r_env = jax_utils.distance_from_one_to_others(...)` is where most of the memory is allocated. I handled this by breaking up the traj into smaller segments, it could be useful to add this as an option to `lj_prefactors_on_traj`. Same with `coulomb_prefactors_on_traj`.

Comment (jkausrelay) [reply timemachine/potentials/nonbonded.py]: 256 should be good (I've been using 200)

**Review by jkausrelay — APPROVED** [review]

File: timemachine/potentials/nonbonded.py (hunk 1 above)
Comment (maxentile) [reply timemachine/potentials/nonbonded.py]: Reduced to 200 to match https://github.com/proteneer/timemachine/pull/931/commits/27d0dc9711f7025980c4d907712e795b16b7bf9f

File: tests/test_jax_nonbonded.py [hunk 2]
```diff
@@ -334,7 +337,38 @@ def u_b(x, box, params):
...
+    np.testing.assert_allclose(u_a(conf, box, params), u_b(conf, box, params))
+
+
+def test_lj_basis():
+    """Randomized test that LJ(sig, eps) computed via basis expansion matches reference"""
+
+    np.random.seed(2023)
+
+    n_env = 10_000
+    r_i = np.linspace(1.9, 8.0, n_env)
```
Comment (mcwitt) [inline tests/test_jax_nonbonded.py:349]: nit: comment on why these values were chosen?

File: timemachine/potentials/nonbonded.py (hunk 1 above)
Comment (mcwitt) [reply timemachine/potentials/nonbonded.py]: I might be missing something but it's not clear to me why `jax_utils.distance_from_one_to_others` should be allocating a lot of memory -- seems like it should just be linear in `x_others`?

File: timemachine/potentials/nonbonded.py [hunk 3]
```diff
@@ -483,51 +490,133 @@ def coulomb_interaction_group_energy(q_ligand: Array, q_prefactors: Array) -> fl
...
+    h_n : [20] array
+        can be dotted with output of project_lj(sig, eps)
+        to compute energy of one atom interacting with all environment atoms
+    """
+    h_n_12 = _basis_expand_lj_term(sig_env, eps_env, r_env, 12)
+    h_n_6 = -_basis_expand_lj_term(sig_env, eps_env, r_env, 6)
+    return jnp.hstack([h_n_12, h_n_6])
+
+
+def project_lj(sig: float, eps: float) -> Array:
```
Comment (mcwitt) [inline timemachine/potentials/nonbonded.py:560]: nit: initially I was a bit thrown off by the naming here, since I'm used to thinking of a projection as something that discards information, but I think this mapping is invertible? Is there another reason to call it a projection?

File: tests/test_jax_utils.py [hunk 4]
```diff
@@ -181,3 +182,23 @@ def test_pairwise_distances_periodic_lifting(coords_box_w):
...
+    traj = np.random.randn(T, N, 3)
+    boxes = np.random.randn(T, 3, 3)
+
+    def f_snapshot(x, box):
+        return jnp.sum(x ** 2) + jnp.sum(box ** 3)  # arbitrary fxn
+
+    reference = vmap(f_snapshot)(traj, boxes)
+    actual = process_traj_in_chunks(f_snapshot, traj, boxes)
+
+    np.testing.assert_allclose(actual, reference)
```
Comment (mcwitt) [inline tests/test_jax_utils.py:204]: Should this test different `chunk_size`s? (including edge cases `chunk_size == 1`, `chunk_size == T`, and `T % chunk_size != 0`)

File: tests/test_jax_nonbonded.py (hunk 2 above)
Comment (maxentile) [reply tests/test_jax_nonbonded.py:349]: Willfix.

(This range was inherited from an initial plot using combined_sigma = fixed_value_near_1 + variable_value_near_1`. The lower end of the range was chosen to be just under 2 to avoid too much clashing, and the upper end of the range was chosen arbitrary and large.)

File: timemachine/potentials/nonbonded.py (hunk 3 above)
Comment (maxentile) [reply timemachine/potentials/nonbonded.py:560]: This name is not well thought-out -- happy to change to any preferred alternative.

Comment (mcwitt) [reply timemachine/potentials/nonbonded.py:560]: Hmm, would something like

- `basis_expand_lj` -> `basis_expand_lj_env`
- `project_lj` -> `basis_expand_lj_atom`

make sense?

File: tests/test_jax_utils.py (hunk 4 above)
Comment (maxentile) [reply tests/test_jax_utils.py]: Addressed in https://github.com/proteneer/timemachine/pull/931/commits/bc584186a026e5859070a70c1d3291bff0f79baf

Comment (mcwitt) [reply tests/test_jax_utils.py]: Could this assert equality?

File: timemachine/potentials/nonbonded.py (hunk 3 above)
Comment (maxentile) [reply timemachine/potentials/nonbonded.py]: Sounds good to me -- applied in https://github.com/proteneer/timemachine/pull/931/commits/8dc364cf0a1d01eaf9230afff773c6795b998619

File: tests/test_jax_utils.py (hunk 4 above)
Comment (maxentile) [reply tests/test_jax_utils.py]: Yep, strengthened in https://github.com/proteneer/timemachine/pull/931/commits/f06fee7ce1a1f4730c820032350275fb7ee54410

**Review by mcwitt — APPROVED** [review]
Looks great to me!

I'm still a little confused about why batching helps significantly with memory usage, but it doesn't seem like it would hurt and I don't think this should block merging.

File: timemachine/potentials/nonbonded.py (hunk 1 above)
Comment (maxentile) [reply timemachine/potentials/nonbonded.py]: I think the size of this intermediate array may be something like `traj_length * n_ligand_atoms * n_environment_atoms * 3` (but I'm not sure). Although linear in `traj_length`, the constant factor is probably large.

Follow-ups could be:
* to provide Jax some clearer hint to "fuse" the computation distance matrix with the sum reduction / avoid this large intermediate array
* to write a "custom op" implementation that uses the neighborlist

I'm leaning towards just doing the second option, since
(1) there would also be utility in computing these on the fly (to avoid ever having to store trajectories),
(2) I don't foresee a need to use any other jax transformations with these preprocessing functions (I don't see a compelling reason to keep the implementation in Jax if doing so is inconvenient / introduces surprising resource requirements)
